qiskit/mapper/_layout.py

```diff
@@ -0,0 +1,145 @@
+
+    def __getitem__(self, item):
+        if item is None:
+            return None
+        if isinstance(item, int) and item < len(self) and item not in self:
+            return None
+        return dict.__getitem__(self, item)
+
+    def __setitem__(self, key, value):
+        if key in self:
+            del self[key]
+        if value in self:
+            del self[value]
+        dict.__setitem__(self, key, value)
+        dict.__setitem__(self, value, key)
+
+    def __delitem__(self, key):
+        dict.__delitem__(self, self[key])
+        dict.__delitem__(self, key)
+
+    def __len__(self):
+        return max([key for key in self.keys() if isinstance(key, int)], default=-1) + 1
+
+    def add(self, logical, physical=None):
+        """
+        Adds a map element between `logical` and `physical`. If `physical`
+        is not defined, `logical will be mapped to a new wire (extending
+        the length of the layout)
+        Args:
+            logical (tuple): A logical (qu)bit. For example, ('qr',2).
+            physical (int): A physical wire. For example, 3.
+        """
+        if physical is None:
+            physical = len(self)
+        self[logical] = physical
+
+    def length(self, amount_of_wires):
+        """
+        Extends the layout length to `amount_of_wires`.
+        Args:
+            amount_of_wires (int): The amount of wires to set in the layout.
+        Raises:
+            LayoutError: If amount_of_wires is used to reduced the length instead
+                of extending it.
+        """
+        current_length = len(self)
+        if amount_of_wires < current_length:
+            raise LayoutError('Lenght setting cannot be smaller than the current amount of wires.')
+        for new_wire in range(current_length, amount_of_wires):
+            self[new_wire] = None
+
+    def idle_wires(self):
+        """
+        Returns the wires that are not mapped to a logical (qu)bit.
+        """
+        idle_wire_list = []
+        for wire in range(self.__len__()):
+            if self[wire] is None:
+                idle_wire_list.append(wire)
+        return idle_wire_list
+
+    def get_logical(self):
+        """
+        Returns the dictionary where the keys are logical (qu)bits and the
+        values are physical wires.
+        """
+        return {key: value for key, value in self.items() if isinstance(key, tuple)}
+
+    def get_physical(self):
+        """
+        Returns the dictionary where the keys are physical wires and the
+        values are logical (qu)bits.
+        """
+        return {key: value for key, value in self.items() if isinstance(key, int)}
+
+    def swap(self, left, right):
+        """ Swaps the map between left and right.
+        Args:
+            left (tuple or int): Item to swap with right.
+            right (tuple or int): Item to swap with left.
+        """
+        temp = self[left]
+        self[left] = self[right]
+        self[right] = temp
```

Comment:
should this enforce the type of `left` and `right` are similar? what if someone writes `layout.swap((qr, 2), 4)`

Reply:
Agree. Fixed in 50c98f2